src/server/gc_worker/compaction_filter.rs

```diff
@@ -0,0 +1,341 @@
+        if let Some(write_batch) = write_batch {
+            write_batch.delete_cf(CF_WRITE, key).unwrap();
+            return;
+        }
+        self.write_batch.delete_cf(CF_WRITE, key).unwrap();
+        self.flush_pending_writes_if_need();
+    }
+
+    fn flush_pending_writes_if_need(&mut self) {
+        if self.write_batch.data_size() > DEFAULT_DELETE_BATCH_SIZE {
+            let mut opts = WriteOptions::new();
+            opts.set_sync(false);
+            let raw_batch = self.write_batch.as_inner();
+            self.db.write_opt(raw_batch, &opts).unwrap();
+            self.write_batch.clear();
+        }
+    }
+
+    fn reset_statistics(&mut self) {
+        if self.versions != 0 {
+            MVCC_VERSIONS_HISTOGRAM.observe(self.versions as f64);
+            self.versions = 0;
+        }
+        if self.deleted != 0 {
+            GC_DELETE_VERSIONS_HISTOGRAM.observe(self.deleted as f64);
+            self.deleted = 0;
+        }
+    }
+}
+
+impl Drop for WriteCompactionFilter {
+    fn drop(&mut self) {
+        let db = self.db.clone();
+        let mut write_batch = RocksWriteBatch::with_capacity(db, DEFAULT_DELETE_BATCH_SIZE);
+        for (_, seek_key) in std::mem::take(&mut self.leveled_tail_deletes) {
+            // In this compaction, the last MVCC version is deleted. However there could be
+            // still some versions for the key which are not included in this compaction.
+            let cf_handle = get_cf_handle(&self.db, CF_WRITE).unwrap();
+            let mut iter = DBIterator::new_cf(self.db.clone(), cf_handle, ReadOptions::new());
+            let mut valid = iter.seek(SeekKey::Key(&seek_key)).unwrap();
+            while valid {
+                let (key, value) = (iter.key(), iter.value());
+                let key_prefix = match Key::split_on_ts_for(key) {
+                    Ok((prefix, _)) if !seek_key.starts_with(prefix) => break,
+                    Ok((prefix, _)) => prefix,
+                    Err(_) => break,
+                };
+
+                let write = WriteRef::parse(value).unwrap();
+                let (start_ts, short_value) = (write.start_ts, write.short_value);
+                if short_value.is_none() {
+                    let key = Key::from_encoded_slice(key_prefix).append_ts(start_ts);
+                    self.delete_default_key(key.as_encoded(), Some(&mut write_batch));
+                }
+
+                self.delete_write_key(key, Some(&mut write_batch));
+                self.versions += 1;
+                self.deleted += 1;
+                valid = iter.next().unwrap();
+            }
+            self.db.write(write_batch.as_inner()).unwrap();
+        }
+        if !self.write_batch.is_empty() {
+            let mut opts = WriteOptions::new();
+            opts.set_sync(true);
+            let raw_batch = self.write_batch.as_inner();
+            self.db.write_opt(raw_batch, &opts).unwrap();
+        } else {
+            self.db.flush(true).unwrap();
```

Comment:
Why flush defaultcf? I think it is better to call 'sync_wal' because if there are only few keys written into memtable, flush is expensive.